sequencing_report_service/services/local_runner_service.py

```diff
@@ -51,51 +51,53 @@ def __init__(self, job_repo_factory, nextflow_command_generator, nextflow_log_di
     async def _start_process(self, job_id):
         with self._job_repo_factory() as job_repo:
             job = job_repo.get_job(job_id)
-            sys_env = os.environ.copy() or {}
-            job_env = job.environment or {}
-            env = {**sys_env, **job_env}
+            assert job
 
             working_dir = os.path.join(
                 self._nextflow_log_dirs, str(job_id))
             os.mkdir(working_dir)
             nxf_log = os.path.join(working_dir, "nextflow.out")
-            nxf_log_fh = open(nxf_log, "w", encoding="utf-8")
-
+            sys_env = os.environ.copy() or {}
+            job_env = job.environment or {}
+            env = {**sys_env, **job_env}
             cmd = shlex.split(shlex.quote(" ".join(job.command)))
-            log.debug("Will start command %s", cmd)
-            process = Subprocess(
-                cmd,
-                stdout=nxf_log_fh,
-                stderr=nxf_log_fh,
-                env=env,
-                cwd=working_dir,
-                shell=True,
-            )
-
-            job_repo.set_state_of_job(job_id=job.job_id, state=State.STARTED)
-            job_repo.set_pid_of_job(job.job_id, process.pid)
-
-            return_code = await process.wait_for_exit()
-
-            nxf_log_fh.close()
-
-            with open(nxf_log, encoding="utf-8") as log_file:
-                cmd_log = log_file.read()
-            if return_code == 0:
+
+            try:
+                with open(nxf_log, "w", encoding="utf-8") as nxf_log_fh:
+                    log.debug("Will start command %s", cmd)
+                    process = Subprocess(
+                        cmd,
+                        stdout=nxf_log_fh,
+                        stderr=nxf_log_fh,
+                        env=env,
+                        cwd=working_dir,
+                        shell=True,
+                    )
+
+                    job_repo.set_state_of_job(job_id=job.job_id, state=State.STARTED)
+                    job_repo.set_pid_of_job(job.job_id, process.pid)
+
+                    await process.wait_for_exit()
+
+                with open(nxf_log, encoding="utf-8") as log_file:
+                    cmd_log = log_file.read()
+
                 log.info("Successfully completed process: %s", job.command)
                 job_repo.set_state_of_job(
                     job_id=job.job_id,
                     state=State.DONE,
                     cmd_log=cmd_log,
                 )
-            else:
-                log.error(
-                    "Found non-zero exit code: %s for command: %s",
-                    return_code,
-                    job.command,
-                )
+            except subprocess.CalledProcessError:
+                job = job_repo.get_job(job_id)
```

Comment:
Maybe missing something here, but why is the job fetched again here?

Reply:
Good question, that's because `job` is not updated in real time, so if the state of the database changes (e.g. because the job is being cancelled), it won't be reflected here unless we fetch it again.

Reply:
Ahh makes sense :+1: